### PR TITLE
Transactions

### PR DESCRIPTION
**Pull request by ChristopherDavenport**
Adds Transactions Support

### CODE REVIEW

File: core/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala [hunk 1]
```diff
@@ -50,31 +50,29 @@ object RedisConnection{
...
-      Some(
-        input.toList.map(a => Resp.BulkString(Some(a)))
-      )
-    )
-    def withSocket(socket: Socket[F]): F[Resp] = explicitPipelineRequest[F](socket, Chunk.singleton(resp)).map(_.head)
-    connection match {
+  def runRequestInternal[F[_]: Concurrent](connection: RedisConnection[F])(
+    inputs: NonEmptyList[NonEmptyList[String]]): F[F[NonEmptyList[Resp]]] = {
+      val chunk = Chunk.seq(inputs.toList.map(Resp.renderRequest))
+      def withSocket(socket: Socket[F]): F[NonEmptyList[Resp]] = explicitPipelineRequest[F](socket, chunk).flatMap(l => Sync[F].delay(l.toNel.get))
```
Comment (Daenyth) [inline core/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala]: ⛏️ 
```suggestion
      def withSocket(socket: Socket[F]): F[NonEmptyList[Resp]] = explicitPipelineRequest[F](socket, chunk).flatMap(l => Sync[F].delay(l.toNel.getOrElse(sys.error("impossible: empty redis response list")))
```

File: core/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala [hunk 2]
```diff
@@ -86,6 +84,13 @@ object RedisConnection{
     }
   })
 
+  private[rediculous] def closeReturn[F[_]: MonadError[*[_], Throwable], A](fE: F[Either[Resp, A]]): F[A] = 
+    fE.flatMap{
+        case Right(a) => a.pure[F]
+        case Left(e@Resp.Error(_)) => ApplicativeError[F, Throwable].raiseError[A](e)
```
Comment (Daenyth) [inline core/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala:90]: Woof. This is why I like to name my `F` ;)

File: core/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala [hunk 3]
```diff
@@ -86,6 +84,13 @@ object RedisConnection{
...
+        case Left(other) => ApplicativeError[F, Throwable].raiseError[A](new Throwable(s"Rediculous: Incompatible Return Type: Got $other"))
```
Comment (Daenyth) [inline core/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala:91]: ⛏️ As a consumer I really like it when all errors raised from the library inherit a common base type

File: core/src/main/scala/io/chrisdavenport/rediculous/RedisTransaction.scala [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  object RedisTx {
+
+    implicit val m: Monad[RedisTx] = new StackSafeMonad[RedisTx]{
+      def pure[A](a: A): RedisTx[A] = RedisTx(Monad[State[(Int, List[NonEmptyList[String]]), *]].pure(a))
+      def flatMap[A, B](fa: RedisTx[A])(f: A => RedisTx[B]): RedisTx[B] = RedisTx(
+        fa.value.flatMap(f.andThen(_.value))
+      )
+    }
+  }
+  final case class Queued[A](f: List[Resp] => Either[Resp, A])
```
Comment (Daenyth) [inline core/src/main/scala/io/chrisdavenport/rediculous/RedisTransaction.scala:78]: `List` to single `A` feels weird - what's the deal there?

File: examples/src/main/scala/TransactionExample.scala [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+      connection <- RedisConnection.queued[IO](sg, new InetSocketAddress("localhost", 6379), maxQueued = 10000, workers = 2)
+    } yield connection
+
+    r.use {client =>
+      val r = (
+        RedisCommands.ping[RedisTransaction.TxQueued],
+        RedisCommands.get[RedisTransaction.TxQueued]("foo"),
+        RedisCommands.set[RedisTransaction.TxQueued]("foo", "value"),
+        RedisCommands.get[RedisTransaction.TxQueued]("foo")
+      ).tupled
```
Comment (Daenyth) [inline examples/src/main/scala/TransactionExample.scala:27]: 🤔 Is it possible to write transactions in a monadic fashion also? I can see a use case for doing a series of steps that are all in a transaction, where you might want the result of one in the middle - I believe redis supports that, logically.

**Review by Daenyth — COMMENTED** [review]
👍 The usage example looks good. I'd find it a bit easier to consume with some javadoc, but the example is definitely helpful

File: examples/src/main/scala/TransactionExample.scala (hunk 5 above)
Comment (ChristopherDavenport) [reply examples/src/main/scala/TransactionExample.scala:27]: No, it does not support that. All are returned on the EXEC step and flushed as a whole. The transactions you are describing are generally better achieved using the lua scripts api I believe.

File: core/src/main/scala/io/chrisdavenport/rediculous/RedisTransaction.scala (hunk 4 above)
Comment (ChristopherDavenport) [reply core/src/main/scala/io/chrisdavenport/rediculous/RedisTransaction.scala:78]: The entire List of responses is returned on the `EXEC` step. So this is how we extract out our A out of the final list of responses returned. It based on the position of our command in the transaction.

File: core/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala (hunk 3 above)
Comment (ChristopherDavenport) [reply core/src/main/scala/io/chrisdavenport/rediculous/RedisConnection.scala:91]: Yeah. I could make some general type at the top level and have others inherit that. Wouldn't be matchable. I'll do that in follow up PR though as I don't want to confuse this one with the error manipulation

**Pull request merged**